### PR TITLE
fix compile warning on copy ctor of Wasm

### PR DESCRIPTION
**Pull request by silentdai**
*Description*:
Fix the warning message below.
```
source/extensions/common/wasm/wasm.cc:1021:1: error: base class 'class std::enable_shared_from_this<Envoy::Extensions::Common::Wasm::Wasm>' should be explicitly initialized in the copy constructor [-Werror=extra]
 Wasm::Wasm(const Wasm& wasm) {
 ^~~~
```
*Risk Level*: 
*Testing*: on github hook
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue]
[Optional *Deprecated*:]

### CODE REVIEW

**Comment by silentdai** [comment]
CC @PiotrSikora @jplevyak

File: source/extensions/common/wasm/wasm.cc [hunk 1]
```diff
@@ -1018,7 +1018,7 @@ void Wasm::getFunctions() {
 #undef _GET
 }
 
-Wasm::Wasm(const Wasm& wasm) {
+Wasm::Wasm(const Wasm& wasm) : enable_shared_from_this<Wasm>() {
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm.cc]: What compiler do you use and what warning are you getting?

The [`Wasm` class already inherits this](https://github.com/istio/envoy/blob/9dc69fdab263ab9895c787d835b4b5131fafff04/source/extensions/common/wasm/wasm.h#L253), and both constructors are `public`, so this shouldn't be needed, AFAIK.

Comment (PiotrSikora) [reply source/extensions/common/wasm/wasm.cc]: Ah, sorry, I see that you've pasted the error in the opening message.

Does this change look reasonable to you, @jplevyak?

Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: I don't think this is necessary.  By default C++ will use the copy constructor from any public superclass.  Perhaps this is related to the issue you fixed where enabl_shared_from_this was not public?

Comment (silentdai) [reply source/extensions/common/wasm/wasm.cc]: @jplevyak I agree with you that the behavior doesn't change with or without the commit.
It's compiler warning not an error. As this is the only low prior error at my side, I prefer to fix it.
```
[-Werror=extra]
```

**Comment by silentdai** [comment]
Sorry I didn't realize there was conflict. Now I resolve it.

@jplevyak As this PR is not evil and it does resolve a compiling warning, what else need I do so that you may approve?

File: source/extensions/common/wasm/wasm.cc (hunk 1 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: What I am trying to figure out is how you are seeing this.  Did you add -Werror=extra in your build?  I tried to rebuild that file and I am not seeing it.  Could you attach repro instructions.

**Comment by silentdai** [comment]
This PR is not need because I accidentally merged this one into #20 
Still tracking the root cause with @jplevyak 
Maybe because CXX on my side default use clang++4
Trying if set CXX to clang++-7 solve this

**Comment by PiotrSikora** [comment]
> Maybe because CXX on my side default use clang++4
> Trying if set CXX to clang++-7 solve this

Please use clang++-7 (what Envoy is currently using in its builds) or clang++-8.

You're going to run into all kinds of issues with outdated compiler.

**Comment by silentdai** [comment]
resolve conflict
CC @jplevyak

**Comment by silentdai** [comment]
Update: this warning is not because of using older clang. I specify clang++-7 but still reports the same error at my side if this PR is not there.

The warning is gone if using run7.sh at silentdai/envoy:withJohn. Will investigate some other day.

In summary, this PR fixes an legit warning.

**Comment by PiotrSikora** [comment]
wrt `run7.sh` - Envoy has rather complex build system (see: [`bazel`](https://github.com/envoyproxy/envoy/tree/master/bazel) subdirectory), and compiling files directly, instead of using Bazel, won't get you anywhere.

Do you see this error/warning when building with Bazel, or only when compiling files directly with your compiler?

**Comment by silentdai** [comment]
@PiotrSikora Yes, I see the error described in the comment of of PR using bazel. Without this PR, I can reproduce every time when I ran "bazel build source/extensions/common/wasm:wasm_lib". However, @jplevyak doesn't see this error when he runs the same command.

To locate the fundamental divergence, @jplevyak and I use reallysilentidai/envoy:withJohn to sync.
- `bazel build -c opt source/extensions/common/wasm:shared_from_test` is targeted to simplifed the dependency and the cpp code to understand why my bazel reports error. Yet, @jplevyak always builds successfully while I always fail.
- `run7.sh` is targeted to figure out which argument is causing the compling error at my side. I used to suspect `-Wextra` is the trouble maker. However, I find that `-Wextra` is not the only factor. The run7.sh has the `-Wextra` but `run7.sh` succeeds.

**Comment by silentdai** [comment]
@jplevyak Could we merge this even the warning is not reproduciable on each enviroment?

**Comment by silentdai** [comment]
@PiotrSikora @jplevyak mystery revealed: It's gcc that report that warning
https://gist.github.com/silentdai/940bd5512e7be825ec0a42bd1963548c

Somehow cc on my desktop is gcc while cxx is clang++

As [envoy declares to support gcc](https://github.com/envoyproxy/envoy/blob/master/bazel/README.md) so we need this PR

**Comment by jplevyak** [comment]
This is merged.  I am closing it.

**Pull request merged**